zcash_primitives/src/sapling/address.rs

```diff
@@ -0,0 +1,118 @@
+use group::{Group, GroupEncoding};
+
+use super::{
+    keys::Diversifier,
+    note::{Note, Rseed},
+};
+
+/// A Sapling payment address.
+///
+/// # Invariants
+///
+/// `pk_d` is guaranteed to be prime-order (i.e. in the prime-order subgroup of Jubjub,
+/// and not the identity).
+#[derive(Clone, Debug)]
+pub struct PaymentAddress {
+    pk_d: jubjub::SubgroupPoint,
+    diversifier: Diversifier,
+}
+
+impl PartialEq for PaymentAddress {
+    fn eq(&self, other: &Self) -> bool {
+        self.pk_d == other.pk_d && self.diversifier == other.diversifier
+    }
+}
+
+impl Eq for PaymentAddress {}
+
+impl PaymentAddress {
+    /// Constructs a PaymentAddress from a diversifier and a Jubjub point.
+    ///
+    /// Returns None if `pk_d` is the identity.
+    pub fn from_parts(diversifier: Diversifier, pk_d: jubjub::SubgroupPoint) -> Option<Self> {
+        if pk_d.is_identity().into() {
+            None
+        } else {
+            Some(PaymentAddress { pk_d, diversifier })
+        }
+    }
+
+    /// Constructs a PaymentAddress from a diversifier and a Jubjub point.
+    ///
+    /// Only for test code, as this explicitly bypasses the invariant.
+    #[cfg(test)]
+    pub(crate) fn from_parts_unchecked(
+        diversifier: Diversifier,
+        pk_d: jubjub::SubgroupPoint,
+    ) -> Self {
+        PaymentAddress { pk_d, diversifier }
+    }
+
+    /// Parses a PaymentAddress from bytes.
+    pub fn from_bytes(bytes: &[u8; 43]) -> Option<Self> {
+        let diversifier = {
+            let mut tmp = [0; 11];
+            tmp.copy_from_slice(&bytes[0..11]);
+            Diversifier(tmp)
+        };
+        // Check that the diversifier is valid
+        diversifier.g_d()?;
```

Comment:
```suggestion
        // PaymentAddress::from_parts will check that the diversifier is valid.
```
(This depends on the change above, and is just an optimization / avoids redundancy. It does not matter what order the components are checked in.)